pallene/translator.lua

```diff
@@ -46,65 +48,105 @@ function Translator:add_whitespace(input, start_index, stop_index)
     assert(start_index <= stop_index+1)
     self:add_previous(input, start_index - 1)
 
-    local region = input:sub(start_index, stop_index)
+    local region = self.input:sub(start_index, stop_index)
     local partial = region:gsub("%S", " ")
     table.insert(self.partials, partial)
 
     self.last_index = stop_index + 1
 end
 
-function translator.translate(input, prog_ast)
-    local instance = Translator:new()
+function Translator:add_local(start_index, stop_index)
+    self:add_previous(start_index - 1)
+    -- The export keyword is six characters long, whereas the local keyword is five characters
+    -- long. Therefore, we pad the keyword with a space. We could add the space to the left, too.
+    -- But it seems more "natural" on the right.
+    table.insert(self.partials, "local ")
 
-    print(require("inspect")(prog_ast))
+    self.last_index = stop_index
+end
 
-    for _, node in ipairs(prog_ast) do
-        if node._tag == "ast.Toplevel.Var" then
-            for _, decl in ipairs(node.decls) do
-                if decl.type then
-                    -- Remove the colon but retain any adjacent comment to the right.
-                    instance:add_whitespace(input, decl.col_loc.pos, decl.col_loc.pos)
-                    -- Remove the type annotation but exclude the next token.
-                    instance:add_whitespace(input, decl.type.loc.pos, decl.end_loc.pos - 1)
-                end
-            end
-        elseif node._tag == "ast.Toplevel.Func" then
-            -- Remove type annotations from function parameters.
-            for _, arg_decl in ipairs(node.value.arg_decls) do
-                -- Type annotations are mandatory for function parameters.
-                -- Remove the colon but retain any adjacent comment to the right.
-                instance:add_whitespace(input, arg_decl.col_loc.pos, arg_decl.col_loc.pos)
-                -- Remove the type annotation but exclude the next token.
-                instance:add_whitespace(input, arg_decl.type.loc.pos, arg_decl.end_loc.pos - 1)
+function Translator:add_exports()
+    if #self.exports > 0 then
+        table.insert(self.partials, "\nreturn {\n")
+        for _, export in ipairs(self.exports) do
+            local pair = string.format("    %s = %s,\n", export, export)
+            table.insert(self.partials, pair)
+        end
+        table.insert(self.partials, "}\n")
+    end
+end
+
+function Translator:translate_decl(decl)
+    if decl.type then
+        -- Remove the colon but retain any adjacent comment to the right.
+        self:add_whitespace(decl.col_loc.pos, decl.col_loc.pos)
+        -- Remove the type annotation but exclude the next token.
+        self:add_whitespace(decl.type.loc.pos, decl.end_loc.pos - 1)
+    end
+end
+
+function Translator:translate_stat(block)
```

Comment:
Could you make this function similar to the other ones?

There is a big if-else-if for the possible statement types. If it is an ast.Stat.Block then it calls itself recursively.